dataset/src/main/scala/frameless/TypedEncoder.scala

```diff
@@ -288,23 +288,28 @@ object TypedEncoder extends LowPriorityTypedEncoder {
   }
 
   /** Encodes things using injection if there is one defined */
-  implicit def usingInjection[A, B]
-    (implicit
-      inj: Injection[A, B],
-      trb: TypedEncoder[B],
-      cta: ClassTag[A],
-      tta: TypeTag[A]
-    ): TypedEncoder[A] =
+  implicit def usingInjection[A: TypeTag : ClassTag, B: TypeTag]
+    (implicit inj: Injection[A, B], trb: TypedEncoder[B]): TypedEncoder[A] =
       new PrimitiveTypedEncoder[A] {
         def nullable: Boolean = false
         def sourceDataType: DataType = ScalaReflection.dataTypeFor[A]
         def targetDataType: DataType = trb.targetDataType
 
-        def constructorFor(path: Expression): Expression =
-          Invoke(Literal.fromObject(inj), "invert", sourceDataType, Seq(path))
-
-        def extractorFor(path: Expression): Expression =
-          Invoke(Literal.fromObject(inj), "apply", targetDataType, Seq(path))
+        def constructorFor(path: Expression): Expression = {
+          // I have no Idea why this is needed, see discusison on #30.
```

Comment:
For some reason that I still don't understand `trb.constructorFor(path)` works for the one on this exact line: https://github.com/adelbertc/frameless/pull/30/commits/1bfa25226e97cbe93bdfb3aa811aa61bdc0cbdd4#diff-6917aeb0150d0a857fad88f6d2b7bb1aR72. But the old, wrong implementation (`trb.constructor()`) does work for the above, and fails when there is some nesting.

It took me a lot of time to "make it work" (Catalyst errors are really hard to read :/), and my approach to come up with the current solution was extremely brute force.

I first tried to print whatever I could to see if I could discriminate the two cases in a sensible way:

    // Accessing these triggers a side effect which results in an incomprehensible error
    // println(path.dataType)
    // println(path.nullable)

    println()
    println(path)
    println("foldable: " + path.foldable)
    println("deterministic: " + path.deterministic)
    println("references: " + path.references)
    println("childrenResolved: " + path.childrenResolved)
    println("semanticHash: " + path.semanticHash)
    println("checkInputDataTypes: " + path.checkInputDataTypes)
    println("prettyName: " + path.prettyName)
    println("children: " + path.children)
    println("resolved: " + path.resolved)
    println("getClass.getName: " + path.getClass.getName)

But none of this give can be used to distinguish the failing case. Then I found that if you pattern match on `path` you can deconstruct it, which allows to access the `dataType` without triggering the side effect. The failing case seems to correspond to a `dataType` of type `StructType`, hence the current implementation.